feat(person-override): Add a helper model to indirectly reference person

tomasfarias opened the pull request:
## Problem

This allows us to use an exclusion constraint on the person overrides table instead of directly using a FK on `posthog_person`. Basically, the helper model acts as a mapping from a serial ID to a person UUID. This will break if we ever have more than 2147483647 overrides. The solution is that the squash erases old ids (erases all of them after squashing) and "resets the PK".

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

Adds a new model, tests. Should the model be called `PersonOverridesMapping` instead?

EDIT: Model has been renamed to `PersonOverridesMapping`.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## How did you test this code?

Likely unit tests will not work until we modify merge transaction. Afterwards we should be able to run the same test suite without modifications.

Although this migration fails the safety checks it should be safe:

* We are deleting fields from a model that is not currently in use (`PersonOverride`) and creating constraints only in models that are not in use (`PersonOverride` and `PersonOverrideHelper`). We are not cleaning up the FK in `PersonOverride` referencing ~~`Person`~~ `Team` to avoid exclusive locks, even though it's not necessary anymore. 

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

tomasfarias committed:
feat(person-override): Add a helper model to indirectly reference person overrides

This allows us to use an exclusion constraint on the person overrides table instead of directly using a FK on posthog_person.

tomasfarias committed:
test(person-overrides): Update tests to match new constraint

tomasfarias commented on posthog/migrations/0306_add_indirect_person_override_constraints.py:
```diff
@@ -0,0 +1,96 @@
+DROP_FUNCTION_FOR_CONSTRAINT_SQL = "DROP FUNCTION is_override_person_not_used_as_old_person"
+
+
+class Migration(migrations.Migration):
```
This is a thick migration. But unsafe operations are limited to models that are not in use currently (`PersonOverrides` mostly).

Notice we are not dropping the `team` FK from `PersonOverrides` even though it's redundant. This is to avoid table locks on `Team` during migration. `Team`, in contrast to `PersonOverrides` is used a lot so locking is bad.

tomasfarias committed:
fix(migration): Add drop extension query to reverse migration

tomasfarias committed:
fix(migration): Use correct table name

tiina303 commented on posthog/models/test/test_person_override_model.py:
```diff
@@ -2,9 +2,10 @@
 from django.db.utils import IntegrityError
 
-from posthog.models import PersonOverride, Team
+from posthog.models import PersonOverride, PersonOverrideHelper, Team
```
Could you add the extra changes/tests from the reverted PR https://github.com/PostHog/posthog/pull/14277
- use different team/org
```
organization = create_organization(name="test")
team = create_team(organization=organization)
```
instead of 
```
PersonOverride.objects.all().delete()
PersonOverrideHelper.objects.all().delete()
```
- test_person_override_disallows_overriding_to_itself - Test old_person_id cannot match override_person_id
- test_person_override_disallows_concurrent_merge

tomasfarias replied:
`create_team` and `create_organization` do not delete state after test is finished. I would use the fixtures in #14513

tiina303 replied:
and the fact that they don't is great - makes debugging much easier as you can inspect the db state after the test

tomasfarias replied:
But that creates inconsistent state for each test.

tomasfarias replied:
I've brought over the new tests from #14227 (albeit in their updated form as seen in #14513), as well as the comment edit pointed out in #14513.

Locally they are passing, but waiting for CI to be green too.

hazzadous commented on posthog/migrations/0307_add_indirect_person_override_constraints.py:
```diff
@@ -0,0 +1,96 @@
+        migrations.AddField(
+            model_name="personoverridehelper",
+            name="team",
+            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to="posthog.team"),
```
You're ok with adding these as ForeignKey? Might make things easier to play with/less risky with side-effects of e.g. deleting a team if it was a weaker link.

tomasfarias replied:
Don't feel super strongly about this constraint, so I'm fine with dropping it. The same one exists on `personoverrides` for now anyways, so we'll catch any unknown teams, but that could also be removed eventually.

hazzadous reviewed (approved):
Looks good.

Just a small comment regarding team_id REFERENCES posthog_team table and if we need that strong a constraint. But happy to go with it if it's useful so non-blocking

tiina303 reviewed (commented):
looks fine to me, but I'm no django expert

in the description you had
`We are not cleaning up the FK in PersonOverride referencing Person to avoid exclusive locks, even though it's not necessary anymore.` but it should be `referencing Team`, right?

tomasfarias committed:
refactor(person-overrides): Make team a regular bigint field without FK

tomasfarias commented on posthog/models/person/person.py:
```diff
@@ -123,50 +134,25 @@ class Meta:
     id = models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name="ID")
     team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
 
     # We don't want to delete rows before we had a chance to propagate updates to the events table.
     # To reduce potential side-effects, these are not ForeingKeys.
```
Yup, I meant team, whoops. I'll fix this comment.

tomasfarias replied:
Ah, there is nothing to fix here.

tiina303 replied:
the comment says these are not FKs but they are clearly

tomasfarias replied:
Ah yeah, the comment is out of date. I thought this was referencing the previous message as it had no comment.

tomasfarias commented on posthog/models/person/person.py:
```diff
@@ -104,6 +102,19 @@ class Meta:
+            models.UniqueConstraint(fields=["team", "uuid"], name="unique_uuid"),
+        ]
+
+    id = models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name="ID")
```
Keep an eye on it :sweat_smile:

tomasfarias replied:
Expanding on the idea: query `max(id)` every once in a while, raise an alarm when we get close.

tiina303 replied:
^ yup exactly was thinking of that - obv doesn't need to be in this PR, but we should not forget it.

tomasfarias replied:
The important part to keep in mind is that the squash job _must_ "reset" the serial PK, otherwise we would eventually always reach the limit.

tomasfarias committed:
refactor(person-overrides): Rename Helper model to Mapping

tomasfarias committed:
chore: Remove out-of-date comment

tomasfarias committed:
refactor(person-overrides): Incorporate concurrency tests

This commit adds new tests that were added in a separate branch to
keep changes to the model contained in a single branch.

These changes include a set of new concurrency-related tests as well
as some minor comment editing.

ellie commented on posthog/migrations/0307_add_indirect_person_override_constraints.py:
```diff
@@ -0,0 +1,92 @@
+            ),
+        ),
+        # Provides operator classes for integers (gist_int4_ops)
+        migrations.RunSQL("CREATE EXTENSION btree_gist", "DROP EXTENSION btree_gist"),
```
are these extensions all standard? ie in a standard postgres distribution

ellie replied:
All good ✅ 

https://www.postgresql.org/docs/current/btree-gist.html

tomasfarias replied:
They are in the PG distribution since at least version 9.6: https://www.postgresql.org/docs/9.6/contrib.html

tomasfarias replied:
However, it can depend on our specific deployment:
> If you are using a pre-packaged version of PostgreSQL, these modules are typically made available as a separate subpackage, such as postgresql-contrib.
